Docs: Concurrency Check & `IHasConcurrencyStamp` Interface (#11981)
abpframework/abp: docs/en/Concurrency-Check.md

## Concurrency Check

### Introduction

Concurrency Check (also known as **Concurrency Control**) refers to specific mechanisms used to ensure data consistency in presence of concurrent changes (multiple processes or users access or change the same data in a database at the same time).

There are two commonly used concurrency control mechanisms/approaches;

> * **Optimistic Locking**: Optimistic Locking is a concurrency control method that does not use record locking. Instead, allows multiple users to attempt to **update** or **delete** the same record without informing the users that others are also attempting to **update/delete** the record. 

Reviewer: `Optimistic Locking` -> `Optimistic concurrency control`. There is no locking at all.

    * If a user successfully updates/deletes the record, the other users need to get the latest changes for the current record to be able to make changes. If any other user attempts to change the record without getting the latest state of it, informed that a conflict exists. 

> * **Pessimistic Locking**: Pessimistic Locking prevents simultaneous updates to records. As soon as one user starts to edit a record, it is getting locked. Other users who attempt to update this record are informed that another user has an update in progress. The other users must wait until the first user has finished committing their changes, thereby releasing the record lock. Only then other users can make changes to this record.

Reviewer: `Pessimistic locking` -> `Pessimistic concurrency control`

> 

Reviewer: Right here, we should say that ABP's concurrency check system uses the optimistic concurrency. Otherwise, user may think ABP provides both.

Also, the explanations in this section is general. it would be good if you give some reference links to these topics for who wants to learn more.

> ### Base Classes & Interfaces for Concurrency Stamp

Reviewer: Rename this title as "Usage"

#### `IHasConcurrencyStamp` Interface

> It is the base interface for **Concurrency Check** and has a just simple property named `ConcurrencyStamp`. 

Reviewer: Move this to under the interface definition, and give more information how ABP manages this property.

```csharp
public interface IHasConcurrencyStamp 
{
    public string ConcurrencyStamp { get; set; }
}
```

> To be able to use **Concurrency Check** your entity class should inherit from the `IHasConcurrencyStamp` interface directly or indirectly.

Reviewer: ```suggestion
To enable concurrency control to your entity class, you should implement the `IHasConcurrencyStamp` interface, directly or indirectly.
```

In general, interfaces are implemented, not inherited (by classes). Inheriting an interface means you create another interface that extend the first interface.

Also, this sentence should be first under the _`IHasConcurrencyStamp` Interface_ header.

#### Base Classes

> Most of the **Entity Classes** provided by ABP Framework, inherit from the `IHasConcurrencyStamp` interface. You can see the full list of the base classes:

Reviewer: Aggregate Root entity classes already implement the `IHasConcurrencyStamp` interface. So, if you are deriving from one of these base classes, you don't need to manually implement the `IHasConcurrencyStamp` interface":

- `AggregateRoot`, `AggregateRoot<TKey>`
- `CreationAuditedAggregateRoot`, `CreationAuditedAggregateRoot<TKey>`
- `AuditedAggregateRoot`, `AuditedAggregateRoot<TKey>`
- `FullAuditedAggregateRoot`, `FullAuditedAggregateRoot<TKey>`

> ### Getting Started 

Reviewer: Remove this section, it doesn't add any value.

This is a core feature and it's used by the ABP Framework. There are no additional packages required. **Optimistic Locking** is applied for ABP modules by default.

> #### Usage

Reviewer: Remove this section and move examples to corresponding sub-sections written above.

{
    //...

    public string ConcurrencyStamp { get; set; }
}

public class UpdateBookDto : IHasConcurrencyStamp 
{
    //...

    public string ConcurrencyStamp { get; set; }
}
```

Also, you need to set the **ConcurrencyStamp** to the entity in the **UpdateAsync** method of your application service, for that purpose you can use the `SetConcurrencyStampIfNotNull` method like as below:

```csharp
public class BookAppService : ApplicationService, IBookAppService 
{
    //...

    public virtual async Task<BookDto> UpdateAsync(Guid id, UpdateBookDto input) 
    {
        var book = await BookRepository.GetAsync(id);

        book.SetConcurrencyStampIfNotNull(input.ConcurrencyStamp);

        //set other input values to entity ...

        await BookRepository.UpdateAsync(book);
    }
}
```

After that, when multiple users try to change the same record at the same time, concurrency stamp mismatch occurs and `AbpDbConcurrencyException` will be thrown. You can either handle the exception manually or let the ABP Framework handle it for you. 

ABP Framework shows a user-friendly error message like in the image below, if you don't handle the exception manually.

> 

Reviewer: A better ABP error message can be:

Optimistic concurrency check has been failed. The entity you're working on has modified by another user. Please discard your changes and try again.
